File: 0221-maximal-square/0221-maximal-square.cpp
class Solution {
public:
     int rec(vector<vector<char>>& matrix,int i,int j,int n,int m,vector<vector<int>> &dp,int &maxi)
    {
        if(i>=n || j>=m || matrix[i][j]=='0')
        return 0;

        if(dp[i][j]!=-1)
        return dp[i][j];

        int right=rec(matrix,i,j+1,n,m,dp,maxi);
        int down=rec(matrix,i+1,j,n,m,dp,maxi);
        int diagonal=rec(matrix,i+1,j+1,n,m,dp,maxi);
        if(matrix[i][j]=='1')
        {
            dp[i][j]=1+min({right,down,diagonal});
            maxi=max(maxi,dp[i][j]);
            return dp[i][j];
        }
        else
        return dp[i][j]=0;

    }

    int maximalSquare(vector<vector<char>>& matrix) {
        int n=matrix.size();
        int m=matrix[0].size();
        vector<vector<int>> dp(n,vector<int>(m,-1));
        int maxi=0;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                rec(matrix, i, j, n, m, dp, maxi);
            }
        }
        return maxi*maxi;
    }
};

File: 0221-maximal-square/0221-maximal-square.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        int n=matrix.length;
        int m=matrix[0].length;
        int[][] dp=new int[n][m];
        int maxi=0;
        for(int i=0;i<n;i++)
        {
            dp[i][0]=matrix[i][0]-'0';
            maxi=Math.max(maxi,dp[i][0]);
        }
        for(int i=0;i<m;i++)
        {
            dp[0][i]=matrix[0][i]-'0';
            maxi=Math.max(maxi,dp[0][i]);
        }
        for(int i=1;i<n;i++)
        {
            for(int j=1;j<m;j++)
            {
                if(matrix[i][j]=='1')
                {
                   dp[i][j] = 1 + Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]), dp[i - 1][j - 1]);
                    maxi=Math.max(maxi,dp[i][j]);
                }
                else
                dp[i][j]=0;
            }
        }
        return maxi*maxi;
    }
}